bundles/org.eclipse.cdt.lsp/src/org/eclipse/cdt/lsp/internal/server/CLanguageServerEnableCache.java

```diff
@@ -0,0 +1,198 @@
+	private static final String C_SOURCE = "org.eclipse.cdt.core.cSource"; //$NON-NLS-1$
+	private static final String CXX_SOURCE = "org.eclipse.cdt.core.cxxSource"; //$NON-NLS-1$
+	private static final String C_HEADER = "org.eclipse.cdt.core.cHeader"; //$NON-NLS-1$
+	private static final String CXX_HEADER = "org.eclipse.cdt.core.cxxHeader"; //$NON-NLS-1$
+	private static final Map<URI, Data> cache = Collections.synchronizedMap(new LRUCache<>(BUFFER_SIZE));
+	private static CLanguageServerEnableCache instance = null;
+
+	private CLanguageServerEnableCache() {
+		ContentTypeManager.getInstance().addContentTypeChangeListener(this);
+		if (PlatformUI.isWorkbenchRunning()) {
+			var workbench = PlatformUI.getWorkbench();
+			workbench.addWindowListener(this);
+			Arrays.stream(workbench.getWorkbenchWindows()).map(IWorkbenchWindow::getPages).flatMap(Arrays::stream)
+					.forEach(p -> p.addPartListener(this));
+		}
+	}
+
+	private static void clearAll() {
+		cache.clear();
+	}
+
+	public static void stop() {
+		if (instance != null) {
+			ContentTypeManager.getInstance().removeContentTypeChangeListener(instance);
+			var workbench = PlatformUI.getWorkbench();
+			workbench.removeWindowListener(instance);
+			Arrays.stream(workbench.getWorkbenchWindows()).map(IWorkbenchWindow::getPages).flatMap(Arrays::stream)
+					.forEach(p -> p.removePartListener(instance));
+		}
+		clearAll();
+	}
+
+	public static void clear() {
+		clearAll();
+	}
+
+	public static synchronized CLanguageServerEnableCache getInstance() {
+		if (instance == null) {
+			instance = new CLanguageServerEnableCache();
+		}
+		return instance;
+	}
+
+	public Boolean get(URI uri) {
+		var data = cache.get(uri);
+		return data != null ? data.enable : null;
+	}
+
+	public void put(URI uri, boolean value) {
+		cache.put(uri, new Data(value));
+	}
+
+	@Override
+	public void preferenceChange(PreferenceChangeEvent event) {
+		if (EditorMetadata.PREFER_LSP_KEY.contentEquals(event.getKey())) {
+			clearAll();
+		}
+	}
+
+	@Override
+	public void contentTypeChanged(ContentTypeChangeEvent event) {
+		var id = event.getContentType().getId();
+		if (C_SOURCE.contentEquals(id) || CXX_SOURCE.contentEquals(id) || C_HEADER.contentEquals(id)
+				|| CXX_HEADER.contentEquals(id)) {
+			clearAll();
+		}
+	}
+
+	@Override
+	public void partActivated(IWorkbenchPart part) {
+		// do nothing
+	}
+
+	@Override
+	public void partBroughtToTop(IWorkbenchPart part) {
+		// do nothing
+	}
+
+	// remove cache only if the URI is not opened in any other LSP based editor editor.
+	@Override
+	public void partClosed(IWorkbenchPart part) {
+		if (part instanceof ExtensionBasedTextEditor editor && LspUtils.checkForCContentType(editor.getEditorInput())) {
+			Optional.ofNullable(LSPEclipseUtils.toUri(editor.getEditorInput())).ifPresent(uri -> {
+				var data = cache.get(uri);
+				if (data != null) {
+					if (--data.counter <= 0) {
+						cache.remove(uri);
+					} else {
+						cache.put(uri, data);
```

Comment:
This line seems to be a no-op, just putting back in the same instance as you extracted on line 146